### PR TITLE
Improve tracking of queue's workers and task_queues

### LINKED ISSUE #4304: Workers not being recorded as present in the queue's worker tracking
It's currently 2021-01-25T21:08:02.851Z, so this host is in the past, yet https://firefox-ci-tc.services.mozilla.com/tasks/IjEVDjbhRMaxDSicppecEg was completed 72m ago and doesn't appear in the list of recent tasks here, either.
```
taskcluster=> select * from queue_workers where worker_id = 't-linux64-ms-053';
-[ RECORD 1 ]----+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
worker_group     | mdc1
worker_id        | t-linux64-ms-053
recent_tasks     | [{"runId": 0, "taskId": "HpL2170jSquFLR5xOmyf6g"}, {"runId": 0, "taskId": "PvONiCCZRcCTRjJr5GEG6Q"}, {"runId": 0, "taskId": "GfUfpc_GSzaE8SQZii3XeQ"}, {"runId": 0, "taskId": "Qbr5LAWURbaUQ2H1tY6j4g"}, {"runId": 0, "taskId": "Y-4c79SeRByoLGNuIxkVAQ"}, {"runId": 0, "taskId": "Q6497TduSlK9XpO3AgETbw"}, {"runId": 0, "taskId": "Fp4IZCkaSd2tbt3BMN4dCw"}, {"runId": 0, "taskId": "YR8TGa0mSiepKuoG9m6ing"}, {"runId": 0, "taskId": "Zq9YElg4RByOHGT7F39ZkQ"}, {"runId": 0, "taskId": "KY8sAvIWSzeYr3VhLfkE0A"}, {"runId": 0, "taskId": "ZWWPZfliRKuocImlNU2Eiw"}, {"runId": 1, "taskId": "XKJgtzASS2ueFhnk3mSVKg"}, {"runId": 2, "taskId": "V7ASapmaS8udpKrh5Dwtyg"}]
quarantine_until | 2021-01-23 04:26:36.732+00
expires          | 2021-01-25 02:52:26.886+00
first_claim      | 2021-01-23 04:26:36.732+00
task_queue_id    | releng-hardware/gecko-t-linux-talos-1804
```

### PR DESCRIPTION
**Pull request by djmitche**
Three distinct improvements in three commits, all rolled into a single DB transaction.

Github Bug/Issue: Fixes #4304

### CODE REVIEW

File: db/test/fns/queue_test.js [hunk 1]
```diff
@@ -1159,6 +1179,78 @@ suite(testing.suiteName(), function() {
...
+      assert.equal(res[0].worker_group, 'wg');
+      assert(res[0].quarantine_until < new Date()); // defaults to in the past
+      assert.deepEqual(res[0].expires, expires);
+      helper.assertDateApproximately(res[0].first_claim, new Date());
+      assert.deepEqual(res[0].recent_tasks, []);
+    });
+
+    helper.dbTest('queue_worker_seen updates rows', async function(db) {
+      // only the expires field gets updated (#4366 would add last_active_date as well)
+      const expireses = [
```
Comment (jwhitlock) [inline db/test/fns/queue_test.js:1210]: expireses as at the plural of expires... 🤦

File: db/test/fns/queue_test.js [hunk 2]
```diff
@@ -1044,42 +1044,67 @@ suite(testing.suiteName(), function() {
...
+        JSON.stringify([{ taskId: 'recent', runId: 0 }]),
+      );
       const res = await db.fns.get_queue_worker_tqid('prov/wt', 'wg', 'wi', new Date());
       assert.equal(res[0].task_queue_id, 'prov/wt');
       assert.equal(res[0].worker_group, 'wg');
       assert.deepEqual(res[0].quarantine_until, quarantineUntil);
       assert.deepEqual(res[0].expires, expires);
       assert.deepEqual(res[0].first_claim, firstClaim);
-      assert.deepEqual(res[0].recent_tasks, [{ recent: "task" }]);
+      assert.deepEqual(res[0].recent_tasks, [{ taskId: 'recent', runId: 0 }]);
```
Comment (jwhitlock) [inline db/test/fns/queue_test.js:1095]: Is the change in the value of ``.recent_tasks`` going to cause an issue when reading existing data?

File: db/versions/0064.yml [hunk 3]
```diff
@@ -0,0 +1,124 @@
+version: 64
+description: Add `queue_worker_seen` and `queue_task_queue_seen` functions
+methods:
+  create_task_queue: {deprecated: true}
+  update_task_queue: {deprecated: true}
+  task_queue_seen:
+    description: |-
+      Recognize that a task queue has been seen, creating it if necessary, updating
+      its properties if not null, and in any case bumping its last seen time time. 
+      The expiration time is not allowed to move backward.
```
Comment (jwhitlock) [inline db/versions/0064.yml:10]: Is a test needed that setting an earlier expiration time leaves it at the later value?

**Review by jwhitlock — APPROVED** [review]
Thanks @djmitche! The code looks complete, does similar work in a safer manner, and the tests and docs are excellent.

File: db/test/fns/queue_test.js (hunk 1 above)
Comment (djmitche) [reply db/test/fns/queue_test.js:1210]: I thought you'd like that ;)

File: db/test/fns/queue_test.js (hunk 2 above)
Comment (djmitche) [reply db/test/fns/queue_test.js:1095]: The old value was just used in the tests; the `{taskId, runId}` format is what's used in production.  At one point in development of this PR I was passing `task_id_in` and `run_id_in` to the DB function, so I had to change this test data to match the correct format.  I subsequently changed that (to just passing `task_in`), but having test data in an accurate format is still a good idea.

File: db/versions/0064.yml (hunk 3 above)
Comment (djmitche) [reply db/versions/0064.yml:10]: Good point --I'll add one before pushing.

Comment (djmitche) [reply db/versions/0064.yml:10]: Ah, this already exists, as a smaller part of a (admittedly fairly long, non-unit-y) larger test:
```
      await taskQueueSeen({
        // try to set expires backward                                                                                            
        expires: expires1,
        description: null,
        stability: null,
      });  
```

**Comment by djmitche** [comment]
(I want to release v41 before landing this)

**Pull request merged**